var/spack/repos/builtin/packages/resolve/package.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class Resolve(CMakePackage, CudaPackage, ROCmPackage):
+    """ReSolve is a library of GPU-resident sparse linear solvers. It contains iterative and direct
+    solvers designed to run on NVIDIA and AMD GPUs, as well as CPU devices."""
+
+    homepage = "https://github.com/ORNL/ReSolve"
+    url = "https://github.com/ORNL/ReSolve.git"
+    maintainers("cameronrutherford", "pelesh", "ryandanehy", "kswirydo")
+
+    # version("1.0.0", submodules=False, branch=develop)
+    # version("develop", submodules=False, branch=develop)
```

Comment:
Looks like the versions are commented out. Should at-least `spack install resolve@develop` be tested to work?

- needs git = URL
-  branch=develop ->  branch="develop"

Also is `spack install resolve@develop+rocm` expected to work in this PR?

Reply:
Yes. We have CI/CD testing this package.py on Incline (AMD + ROCm 5.6.0 from source), Deception (x86_64 and NVIDIA), Newell/Ascent (ppcle64 and NVIDIA), and also have tested non-GPU builds on all those platforms.

I was honestly getting bugs with a sole develop version without a commit sha, but maybe I avoided debugging that while we are prototyping.

Current usag is just to make sure you run a `spack develop --path=<path> resolve@develop`

Reply:
https://github.com/ornl/ReSolve - see here

Reply:
Well I think if this is to merge to spack - it might be good to support at-least `spack install resolve@develop`

the edits required here are pretty minor.

> I was honestly getting bugs with a sole develop version without a commit sha

I'm not sure why you would need "commit sha" with develop

```
balay@petsc-gpu-02:/scratch/balay/spack.y$ git diff
diff --git a/var/spack/repos/builtin/packages/resolve/package.py b/var/spack/repos/builtin/packages/resolve/package.py
index 45d85122ca..e0855b7921 100644
--- a/var/spack/repos/builtin/packages/resolve/package.py
+++ b/var/spack/repos/builtin/packages/resolve/package.py
@@ -12,10 +12,11 @@ class Resolve(CMakePackage, CudaPackage, ROCmPackage):
 
     homepage = "https://github.com/ORNL/ReSolve"
     url = "https://github.com/ORNL/ReSolve.git"
+    git = "https://github.com/ORNL/ReSolve.git"
+
     maintainers("cameronrutherford", "pelesh", "ryandanehy", "kswirydo")
 
-    # version("1.0.0", submodules=False, branch=develop)
-    # version("develop", submodules=False, branch=develop)
+    version("develop", submodules=False, branch="develop")
 
     variant("klu", default=True, description="Use KLU, AMD and COLAMD Libraries from SuiteSparse")
 
balay@petsc-gpu-02:/scratch/balay/spack.y$ ./bin/spack install resolve@develop
<snip>
==> Installing resolve-develop-ldauxrythfdmbt676jfh7ryjt4nu5fzt [35/35]
==> No binary for resolve-develop-ldauxrythfdmbt676jfh7ryjt4nu5fzt found: installing from source
==> No patches needed for resolve
==> resolve: Executing phase: 'cmake'
==> resolve: Executing phase: 'build'
==> resolve: Executing phase: 'install'
==> resolve: Successfully installed resolve-develop-ldauxrythfdmbt676jfh7ryjt4nu5fzt
  Stage: 1.70s.  Cmake: 0.24s.  Build: 2.06s.  Install: 0.11s.  Post-install: 0.06s.  Total: 4.34s
[+] /scratch/balay/spack.y/opt/spack/linux-ubuntu22.04-zen4/gcc-11.4.0/resolve-develop-ldauxrythfdmbt676jfh7ryjt4nu5fzt
```

Reply:
BTW: I get build errors with `./bin/spack install resolve+rocm amdgpu_target=gfx90a`

[spack-build-out.txt](https://github.com/spack/spack/files/13304166/spack-build-out.txt)

Reply:
@balay, does GCC support `--offload-arch` option?

@cameronrutherford: did we build other packages with HIP and GCC as the host compiler?

Reply:
I don't know of a build we have done using GCC as host compiler. @pelesh suggested using gcc 13 offline, so that seems like your next best bet https://gcc.gnu.org/wiki/Offloading#How_to_try_offloading_enabled_GCC

Reply:
This is the last remaining thread that's open. Since we have build with 5.6.0 as a ROCm requirement, we ended up using clang as gcc wasn't working well. I would recruit @nkoukpaizan to help here building with gcc.

Our recommendation is that since llvm-amdgpu is a part of our spack depency, you should build with clang compilers for ROCm backend. If you want to use gcc, then CPU only / CUDA builds would work best.

We can also make an issue in ReSolve in order to support a gcc build alongside our clang builds.

Reply:
> Our recommendation is that since llvm-amdgpu is a part of our spack depency, you should build with clang compilers for ROCm backend.

My note is in the context of xsdk - as this MR is tagged for xsdk. [and not really a blocker for  merging in].

Generally its best to enable builds that might be compatible with the xsdk collection of pkgs. This might mean - some builds might have reduced functionality. With xsdk most builds don't expose their full functionality anyway due to conflicts with other pkgs (or compilers or other build issues). So if a package has less restrictions - then that might give a better chance of working in the collective xsdk build (to expose more features). 

It does not mean compromising critical requirements - but supporting more portable builds if feasible (perhaps with some minor trade-offs)

But if gcc build is best done with `resove~rocm` even for `xsdk+rocm` - thats what will be in xsdk [until the xsdk team figures out better ways to deal with any additional requirements]

I'll defer the approval of this new pkg to spack to  @haampie , @alalazo as you've previously indicated in a prior comment.

Reply:
Is this a blocking item for the PR? I can take a look at building with gcc, but probably not this week.

BTW, when using Spack to build, even if the base compiler is gcc, a lot of packages do something like `args.append(self.define("CMAKE_CXX_COMPILER", spec["hip"].hipcc))`. It's not a sustainable solution, but it's one way to get it to work before better HIP support from CMAKE and other compilers.

Reply:
> Is this a blocking item for the PR

nope.

Hoping  @haampie  or @alalazo can review [as @cameronrutherford previously indicated] and merge this in.

Reply:
I am going to resolve conversation, as this is ready for review, and shouldn't block merging.